### PR TITLE
fix(auth): Impossibility to sign-in when RandomKeyPassword missing

### PR DESCRIPTION
**Pull request by israx**
#### Checklist

- [ ] PR description included
- [ ] `yarn test` passes
- [ ] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/auth/__tests__/providers/cognito/signInWithSRP.test.ts [hunk 1]
```diff
@@ -56,13 +66,27 @@ function setDeviceKeys() {
 		mockedDeviceMetadata.randomPasswordKey,
 	);
 }
+type DeviceKey = 'deviceKey' | 'deviceGroupKey' | 'randomPasswordKey';
```
Comment (ashika112) [inline packages/auth/__tests__/providers/cognito/signInWithSRP.test.ts:69]: nit: could we do something like ? this way we just define the keys in one place. Bascially opposite of what you are doing

```
const deviceKeys = ['deviceKey', 'deviceGroupKey', 'randomPasswordKey'] as const;

type DeviceKey = typeof deviceKeys[number]
```

File: packages/auth/__tests__/providers/cognito/signInWithSRP.test.ts [hunk 2]
```diff
@@ -163,6 +187,58 @@ describe('signIn API happy path cases', () => {
...
+		test.each(deviceKeys)('respondToAuthChallenge should not include device key in the request if any device key in storage is deleted', async deviceKey => {
+           deleteDeviceKey(deviceKey);
+			await signIn({
+				username: lastAuthUser,
+				password: 'XXXXXXXX',
+			});
+
+			expect(respondToAuthChallengeAuthSpy).toHaveBeenCalledTimes(1);
+			const deviceKeyFromRequest = respondToAuthChallengeAuthSpy.mock.calls[0][1].ChallengeResponses?.DEVICE_KEY
+			expect(deviceKeyFromRequest).toBe(undefined)
```
Comment (ashika112) [inline packages/auth/__tests__/providers/cognito/signInWithSRP.test.ts:237]: curious: wouldn't it be undefined pre this change as well? since u r coalescing to undefined? im asssuming the test is adding just the case now

Comment (israx) [reply packages/auth/__tests__/providers/cognito/signInWithSRP.test.ts:237]: I added a test above that doesn't delete the device keys, and checks for an actual value.

Comment (israx) [reply packages/auth/__tests__/providers/cognito/signInWithSRP.test.ts:237]: updated the description to specify the tests that were added

File: packages/auth/__tests__/providers/cognito/signInWithSRP.test.ts (hunk 1 above)
Comment (israx) [reply packages/auth/__tests__/providers/cognito/signInWithSRP.test.ts:69]: I'd prefer to keep the explicit type. There are only a few instances in the library where we use the `typeof` operator.

**Review by AllanZhengYP — APPROVED** [review]
Looks good to me. Thank you @israx

File: packages/auth/__tests__/providers/cognito/signInWithSRP.test.ts (hunk 2 above)
Comment (ashika112) [reply packages/auth/__tests__/providers/cognito/signInWithSRP.test.ts:237]: I guess im wondering more on the line test case for return is null like when all three are removed.

**Review by ashika112 — APPROVED** [review]

File: packages/auth/__tests__/providers/cognito/signInWithSRP.test.ts (hunk 2 above)
Comment (israx) [reply packages/auth/__tests__/providers/cognito/signInWithSRP.test.ts:237]: ooh I think what you mean. So if any key or all of them are missing then the underlying method to get keys will always return null, as shown bellow. This will prevent the library to add any device keys to the request. Hence the request will return undefined if a parameter is not found.

```
const deviceKeys = await tokenOrchestrator.getDeviceKeys(); // keys || null

if(deviceKeys){
 // add keys to request
}

```
